rfc/0077-zone-aware-replica-reads.md

```diff
@@ -0,0 +1,183 @@
+(AZ).
+
+# Use Cases
+
+## No Preference
+
+This is current behaviour, where the SDK does not take into account server
+groups.
+
+![figure 1: No Prefrence](figures/0077-case-1-no-preference.svg)
+
+## Local Only
+
+In this case, all reads will be done only from the local availability zone. It
+is the cheapest solution, although it reduces chances of getting data from
+replicas.
+
+![figure 2: Local Only](figures/0077-case-2-local-only.svg)
+
+## Local First
+
+This case is a compromise between cost and availability, which trades speed. The
+idea is that the SDK will partition set of `[active, replica1, ...]` nodes into
+two groups: local and non-local. Then it will perform requests to local only,
+and only when neither of nodes will return data, send requests to non-local
+group.
+
+![figure 3: Local First](figures/0077-case-3-local-first.svg)
+
+# Changes
+
+New enumeration should be defined for expressing different strategies:
+
+```
+enum ReadPreference {
+    NO_PREFERENCE,
+    LOCAL_ONLY,
+    LOCAL_FIRST,
+};
+```
+
+Each of the operation should get new option, with default value of
+`ReadPreference::NO_PREFERENCE`.
+
+```
+class GetAllReplicasOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class GetAnyReplicaOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class LookupInAllReplicasOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class LookupInAnyReplicaOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+In all failure cases, when the SDK cannot handle operation, it must return `102
+DocumentUnretrievable` with human-readable explanation whenever it is possible
+that explains the details.
+
+# Open Questions
+
+## Q1
+
+What the SDK should do if the local group does not have any replica configured
+for the key, but the User request `Local-Only` strategy? Consider the following
+cases:
+
+- number of replicas is 1, number of zones is 3. Neither active, nor replica is
+  in the local group.
+- number of replicas is 1, number of zones is 3. Only active node is in the
+  local group.
+- number of replicas is 1, number of zones is 3. Only replica node is in the
+  local group.
+
+Shall we warn users when locality restriction turns request into just GET,
+because only active node is in local group?
+
+## Q2
+
+Do we need to implement management API to create/retrieve server groups and
+change node belonging to particular group?
+
+## Q3
+
+All currently supported server versions allow to configure server groups, but
+only 7.6.1 announce this information in configuration.
+[MB-60835](https://issues.couchbase.com/browse/MB-60835) How the SDK should
+behave with older server? Do we need to fail fast or silently fall back to
+current behavior?
```

Comment:
If we treat the desire to write to a specific server group as a request not a command, it gives us a lot of lee-way to handle situations like this.  If we don't have a config with server groups, we fallback to whatever the operation does normally.

Reply:
I updated response to this question, basically the behavior for old servers is not defined as they do not support Zone-Aware-Replica-Reads in the way we describe it (which requires `serverGroup` markers in the configuration).

I think we should rather document this better to make it clear how to use it, and that whole cluster should have at least 7.6.2 on all nodes, and be balanced correctly.